exec/java-exec/src/main/java/org/apache/drill/exec/physical/resultSet/model/hyper/HyperReaderBuilder.java

```diff
@@ -71,30 +72,26 @@
  * the outer vector accessor.)
  */
 
-public abstract class BaseReaderBuilder extends AbstractReaderBuilder {
+public class HyperReaderBuilder extends ReaderBuilder {
 
-  /**
-   * Read-only row index into the hyper row set with batch and index
-   * values mapping via an SV4.
-   */
-
-  public static class HyperRowIndex extends ReaderIndex {
+  private static final HyperReaderBuilder INSTANCE = new HyperReaderBuilder();
 
-    private final SelectionVector4 sv4;
-
-    public HyperRowIndex(SelectionVector4 sv4) {
-      super(sv4.getCount());
-      this.sv4 = sv4;
-    }
+  private HyperReaderBuilder() { }
 
-    @Override
-    public int offset() {
-      return AccessorUtilities.sv4Index(sv4.get(position));
-    }
+  public static RowSetReaderImpl build(VectorContainer container, TupleMetadata schema, SelectionVector4 sv4) {
+    HyperRowIndex rowIndex = new HyperRowIndex(sv4);
+    return new RowSetReaderImpl(schema, rowIndex,
+        INSTANCE.buildContainerChildren(container, schema));
+  }
 
-    @Override
-    public int hyperVectorIndex( ) {
-      return AccessorUtilities.sv4Batch(sv4.get(position));
+  public static RowSetReaderImpl build(BatchAccessor batch) {
+    VectorContainer container = batch.container();
+    try {
+      return build(container,
+          new HyperSchemaInference().infer(container),
+          batch.selectionVector4());
+    } catch (SchemaChangeException e) {
+      throw new UnsupportedOperationException(e);
```

Comment:
Why is we capture schema change exception, unsupported exception is thrown?

Reply:
Revised and added comment.